panel.py

```diff
@@ -82,126 +84,102 @@
-
-  return True
-
-buffer = ''
-def receiveController():
-  global buffer
-
-  while True:
-    try:
-      data = sock.recv(4096) # Arbitrary / maximum length.
-    except:
-      # If the buffer is empty, we return as not having handled an event.
-      # Otherwise we wait for the rest of the data corresponding to this event.
-      if not buffer:
-        return False
-      else:
-        continue
+  player_number = int(args[0])
+  if player_number is 1:
+    controls = controls[:2]
+  elif player_number is 2:
+    controls = controls[2:]
+  else:
+    # To add more players, just add some more controls to the list above so
+    # that we have more to divvy up.
+    print 'Game only supports two players atm.'
+    return 2
 
-    buffer += data
+  print 'Your controls are: ', json.dumps(controls, indent=2)
 
-    # Note that we may receive multiple events at once.
-    while buffer:
-      raw_event, carriage_return, following_text = buffer.partition('\r')
+  try:
+    # start socket operations
+    remote = (
+        os.getenv('CONTROLLER_IP', 'localhost'),
+        os.getenv('CONTROLLER_PORT', 8000),
+      )
+    c = Client()
+    c.connect(remote)
+
+    c.send('announce', { 'controls': controls })
+
+    # start reading keyboard
+    keyboard_reader = threading.Thread(target=readKeyboard)
+    keyboard_reader.start()
+
+    # main loop
+    while True:
+      # read events from the network
+      event = c.read()
+      if event is not None:
+        if event['message'] == 'display':
+          display = event['data']['display']
+          print '> ' + display
+
+      # handle local events
       try:
-        event = json.loads(raw_event)
+        control_id, state = control_queue.pop(0)
       except:
-        # We didn't receive a complet message.
-        buffer = raw_event
-        break
-
-      if event['message'] == 'display':
-        display = event['data']['display']
-        print '> ' + display
-
-      # Resume processing the text after the message--possibly another event.
-      buffer = following_text
-
-  return True
-
+        pass
+      else:
+        try:
+          control = filter(lambda c: c['id'] == control_id, controls)[0]
+        except:
+          print "no controls with control_id %s" % control_id
+        else:
+          c.send('set-state', { 'id': control_id, 'state': state })
+
+          # HACK(jeff): We need to explicitly reset buttons' state while
+          # playing from the CLI, whereas a physical button's state would
+          # automatically reset when the player lifted their finger off.
+          if ('type' in control) and (control['type'] == 'button'):
+            c.send('set-state', { 'id': control_id, 'state': '0' })
+
+      # Wait for an event to occur.
+      time.sleep(0.1)
+  except KeyboardInterrupt:
+    return 0
+  finally:
+    if c:
+      c.stop()
+
+    should_exit = True
+    # TODO: we cannot do this because we use an un-timed-out raw_input
```

Comment:
a panel project (like mine for instance) is going to end up with several threads, and i'd like to be able to do resets without exiting the entire program. i don't think this is necessary, but it can't hurt -- there's nothing worse than a program that won't exit

Reply:
Oh ok!